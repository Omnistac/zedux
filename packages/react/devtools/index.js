if (process.env.NODE_ENV !== 'development') {
  module.exports = {
    getStateHub: function getStateHub() {},
    StateHub: function StateHub() {
      return null
    },
  }
} else {
  module.exports = require('./dev')
}
