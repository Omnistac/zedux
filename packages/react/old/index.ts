export * from './AppProvider'
export { globalStore } from './global/store'
export * from './createAtom'
export * from './createMolecule'
export * from './types'
export * from './useAtom'
export * from './useMolecule'
export * from './useStore'
