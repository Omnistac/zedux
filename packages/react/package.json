{
  "name": "@zedux/react",
  "version": "0.3.4",
  "description": "Elite state management for React",
  "main": "./dist/es/react/src/index.js",
  "module": "./dist/es/react/src/index.js",
  "es2015": "./dist/es/react/src/index.js",
  "publishConfig": {
    "registry": "https://omnistac.jfrog.io/omnistac/api/npm/npm-local/"
  },
  "typings": "./dist/es/react/src/index.d.ts",
  "sideEffects": false,
  "files": [
    "dist"
  ],
  "scripts": {
    "build-storybook": "build-storybook",
    "build:es": "ttsc --outDir dist/es -p tsconfig.es.json",
    "build:umd:min": "cross-env NODE_ENV=production webpack",
    "build:umd": "cross-env NODE_ENV=development webpack",
    "build": "yarn build:es && yarn build:umd && yarn build:umd:min",
    "dev": "yarn build:es -- --watch",
    "prepublishOnly": "yarn build",
    "storybook": "start-storybook -p 6006",
    "test": "cd ../.. && yarn test packages/react"
  },
  "keywords": [
    "abstraction",
    "atom",
    "atoms",
    "atomic",
    "beautiful",
    "boilerplate",
    "DI",
    "dependency",
    "easy",
    "flux",
    "functional",
    "immutable",
    "injection",
    "machine",
    "management",
    "memoized",
    "performant",
    "react",
    "reducer",
    "redux",
    "simple",
    "state",
    "test",
    "tests",
    "testing",
    "zedux"
  ],
  "author": "Joshua Claunch",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Omnistac/zedux.git"
  },
  "bugs": {
    "url": "https://github.com/Omnistac/zedux/issues"
  },
  "homepage": "https://github.com/Omnistac/zedux#readme",
  "devDependencies": {
    "@testing-library/jest-dom": "^5.11.9",
    "@testing-library/react": "^11.2.5",
    "@types/styled-components": "^5.1.7",
    "core-js": "3",
    "react": "^17.0.1",
    "react-dom": "^17.0.1",
    "react-icons": "^4.1.0",
    "styled-components": "^5.2.1",
    "ttypescript": "^1.5.12",
    "typescript-transform-paths": "^2.2.3"
  },
  "dependencies": {},
  "peerDependencies": {
    "react": ">=17.0.1",
    "react-dom": ">=17.0.1"
  }
}
