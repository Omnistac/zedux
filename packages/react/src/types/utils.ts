import { AtomTemplateBase } from '../classes/templates/AtomTemplateBase'
import { AtomInstance } from '../classes/instances/AtomInstance'
import { AtomInstanceBase } from '../classes/instances/AtomInstanceBase'

export type AnyAtomInstance = AtomInstance<any>
export type AnyAtomInstanceBase = AtomInstanceBase<
  any,
  AtomTemplateBase<any, AtomInstance<any>>
>
export type AnyAtomTemplate = AtomTemplateBase<any, any>
