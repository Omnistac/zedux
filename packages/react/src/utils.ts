import { AnyAtomTemplate, Ecosystem } from '@zedux/atoms'
import React, { createContext } from 'react'

export const ecosystemContext = createContext('@@global')

/**
 * These are copied from packages/atoms/src/utils/general.ts
 *
 * IMPORTANT: keep these in-sync with the ones in the atoms package
 */
export const Explicit = 1
export const External = 2
export const Static = 4
// export const Deferred = 8

export const destroyed = Symbol.for(`@@zedux/destroyed`)

export const getReactContext = (
  ecosystem: Ecosystem,
  atom: AnyAtomTemplate
) => {
  const reactStorage: {
    contexts: WeakMap<typeof createContext, Record<string, React.Context<any>>>
  } = (ecosystem._storage.react ||= {})

  const contexts = (reactStorage.contexts ||= new WeakMap())
  const windowContexts =
    contexts.get(createContext) ||
    (contexts.set(createContext, {}).get(createContext) as Record<
      string,
      React.Context<any>
    >)

  return (windowContexts[atom.key] ||= createContext(
    undefined
  ) as React.Context<any>)
}
