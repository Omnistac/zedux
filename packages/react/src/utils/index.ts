export * from './general'
export * from './types'
