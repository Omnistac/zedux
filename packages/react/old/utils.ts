import { Atom, Scope } from './types'

let idCounter = 0

const generateId = () => {
  idCounter += 1
  return idCounter
}

const stringifyParam = (param: any): string => {
  if (Array.isArray(param)) {
    return `[${param.map(child => stringifyParam(child)).join(',')}]`
  }

  if (param && typeof param === 'object') {
    return `{${Object.entries(param)
      .map(([key, val]) => `${key}: ${stringifyParam(val)}`)
      .sort((a, b) => a.localeCompare(b))
      .join(',')}}`
  }

  return JSON.stringify(param)
}

export const generateAppId = () => `app-${generateId()}`
export const generateImplementationId = () => `im-${generateId()}`
export const generateInstanceId = () => `in-${generateId()}`
export const generateLocalId = () => `lo-${generateId()}`

export const getFullKey = (atom: Atom, params?: any[]) => {
  if (atom.scope === Scope.local) return generateLocalId()

  if (!params?.length) return atom.key

  return `${atom.key}---${stringifyParam(params)}`
}
