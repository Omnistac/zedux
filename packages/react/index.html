<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Zedux Dev</title>
  </head>
  <body>
    <div id="root"></div>
    <script type="module" src="/src/index.ts"></script>
    <script type="module">
      import { atom, useAtomState } from './src/index'
      import { createElement } from 'react'
      import { render } from 'react-dom'

      const greetingAtom = atom('greeting', 'Hello, World!')

      const App = () => {
        const [state, setState] = useAtomState(greetingAtom)

        return createElement('div', null,
          createElement('p', null, state),
          createElement('input', {
            onChange: event => setState(event.currentTarget.value),
            value: state
          })
        )
      }

      render(createElement(App), document.getElementById('root'))
    </script>
  </body>
</html>
