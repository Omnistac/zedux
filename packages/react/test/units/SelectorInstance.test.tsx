import { atom, AtomGetters } from '@zedux/react'
import {
  ecosystem,
  getSelectorNodes,
  snapshotSelectorNodes,
} from '../utils/ecosystem'
import { mockConsole } from '../utils/console'

const atom1 = atom('1', () => 'a', { ttl: 0 })
const selector1 = ({ get }: AtomGetters) => get(atom1) + 'b'
const selector2 = ({ select }: AtomGetters) => select(selector1) + 'c'
const selector3 = ({ select }: AtomGetters) => select(selector2) + 'd'

describe('the SelectorInstance class', () => {
  test('deeply nested selectors get auto-created', () => {
    const instance = ecosystem.getNode(selector3)
    const expected = {
      id: '@@selector-selector3-0',
      p: [],
      t: selector3,
      v: 'abcd',
      w: [],
    }

    expect(instance).toEqual(expect.objectContaining(expected))
    snapshotSelectorNodes()
  })

  test('on() adds and removes dependents', () => {
    const instance2a = ecosystem.getNode(selector2)

    expect(getSelectorNodes()).toEqual({
      '@@selector-selector1-1': expect.any(Object),
      '@@selector-selector2-0': expect.any(Object),
    })

    // trigger cleanup without adding a dependent first
    instance2a.destroy()

    expect(getSelectorNodes()).toEqual({})

    const instance2b = ecosystem.getNode(selector2)
    const instance1b = ecosystem.getNode(selector1)
    const cleanup = instance1b.on(() => {})

    instance2b.destroy() // destroys only selector2

    expect(getSelectorNodes()).toEqual({
      // id # is still 1 'cause the Selector class's `_refBaseKeys` still holds
      // the cached key despite `instance2b`'s destruction above
      '@@selector-selector1-1': expect.any(Object),
    })

    cleanup()

    expect(getSelectorNodes()).toEqual({})
  })

  test('ecosystem.dehydrate("@@selector") returns all cached selectors', () => {
    ecosystem.getNode(selector3)

    expect(ecosystem.dehydrate('@@selector')).toEqual({
      '@@selector-selector1-2': 'ab',
      '@@selector-selector2-1': 'abc',
      '@@selector-selector3-0': 'abcd', // selector3's id is created first
    })
  })

  test('destroy() does nothing if the instance is already destroyed', () => {
    const instance = ecosystem.getNode(selector3)
    instance.destroy()

    expect(instance.l).toBe('Destroyed')
    expect(ecosystem.viewGraph()).toEqual({})

    expect(() => instance.destroy()).not.toThrow()
    expect(() => ecosystem.getNode(selector2).destroy()).not.toThrow()
  })

  test('if the selector has dependents, `destroyCache()` bails out unless `force` is passed', () => {
    jest.useFakeTimers()
    ecosystem.getNode(selector3)
    const instance1 = ecosystem.getNode(selector1)
    const instance2 = ecosystem.getNode(selector2)

    ecosystem.getNode(selector1).destroy() // does nothing
    ecosystem.getNode(selector2).destroy() // does nothing

    expect(instance1.l).toBe('Active')
    expect(instance2.l).toBe('Active')
    expect(ecosystem.dehydrate('@@selector')).toEqual({
      '@@selector-selector1-2': 'ab',
      '@@selector-selector2-1': 'abc',
      '@@selector-selector3-0': 'abcd',
    })

    ecosystem.getNode(selector2, []).destroy(true) // destroys both 1 & 2
    jest.runAllTimers()

    expect(instance1.l).toBe('Destroyed')
    expect(instance2.l).toBe('Destroyed')
    expect(ecosystem.dehydrate('@@selector')).toEqual({
      // ids 2 & 1 - the refs are still cached in the Selector class's
      // `_refBaseKeys`
      '@@selector-selector1-2': 'ab',
      '@@selector-selector2-1': 'abc',
      '@@selector-selector3-0': 'abcd',
    })

    ecosystem.getNode(selector3).destroy()

    expect(ecosystem.dehydrate('@@selector')).toEqual({})
  })

  test('ecosystem.find() returns the first selector instance that matches the passed string', () => {
    ecosystem.getNode(selector3)
    const instance2 = ecosystem.getNode(selector2)

    expect(ecosystem.find('selector2')).toBe(instance2)
  })

  // TODO: address the problems in these two tests
  test.skip('ecosystem.findAll() accepts selector refs or string ids', () => {
    const instance = ecosystem.getNode(selector3)
    const allNodes = ecosystem.findAll()

    expect(Object.keys(allNodes)).toEqual([
      '1',
      '@@selector-selector1-2',
      '@@selector-selector2-1',
      '@@selector-selector3-0', // the id for selector3 is generated first
    ])

    const instances1 = ecosystem.findAll('selector2')

    expect(instances1).toEqual({
      '@@selector-selector2-1': ecosystem.getNode(selector2),
    })

    const instances2 = ecosystem.findAll(selector3)

    expect(instances2).toEqual({
      '@@selector-selector3-0': instance,
    })
  })

  test.skip("ecosystem.findAll() returns an empty object if the selector hasn't been cached", () => {
    ecosystem.getNode(selector2)

    expect(ecosystem.findAll(selector3)).toEqual({})
  })

  test('getNode(instance) returns the passed instance', () => {
    const instance = ecosystem.getNode(selector1)

    expect(ecosystem.getNode(instance)).toBe(instance)
  })

  test('selector errors are caught', () => {
    const mock = mockConsole('error')

    const atom1 = atom('1', 'a')

    // @ts-expect-error accessing undefined property
    const selector1 = ({ get }: AtomGetters) => get(atom1).nothing()

    expect(() => ecosystem.getNode(selector1)).toThrowError(/not a function/i)

    expect(mock).toHaveBeenCalledTimes(1)
    expect(mock).toHaveBeenCalledWith(
      expect.stringContaining('encountered an error while running selector'),
      expect.any(TypeError)
    )
  })

  test('different arg-accepting selectors with the exact same name and args create different instances', () => {
    const selector1 = function commonName(_: AtomGetters, id: string) {
      return `${id}b`
    }

    const selector2 = function commonName(_: AtomGetters, id: string) {
      return `${id}c`
    }

    ecosystem.getNode(selector1, ['a'])
    ecosystem.getNode(selector2, ['a'])

    snapshotSelectorNodes()
  })
})
