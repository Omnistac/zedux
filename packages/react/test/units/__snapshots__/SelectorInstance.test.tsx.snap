// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the SelectorInstance class deeply nested selectors get auto-created 1`] = `
{
  "@@selector-selector1-2": {
    "className": "SelectorInstance",
    "observers": {
      "@@selector-selector2-1": {
        "flags": 1,
        "operation": "select",
        "p": undefined,
      },
    },
    "sources": {
      "1": {
        "flags": 1,
        "operation": "get",
        "p": undefined,
      },
    },
    "state": "ab",
    "status": "Active",
    "weight": 2,
  },
  "@@selector-selector2-1": {
    "className": "SelectorInstance",
    "observers": {
      "@@selector-selector3-0": {
        "flags": 1,
        "operation": "select",
        "p": undefined,
      },
    },
    "sources": {
      "@@selector-selector1-2": {
        "flags": 1,
        "operation": "select",
        "p": undefined,
      },
    },
    "state": "abc",
    "status": "Active",
    "weight": 3,
  },
  "@@selector-selector3-0": {
    "className": "SelectorInstance",
    "observers": {},
    "sources": {
      "@@selector-selector2-1": {
        "flags": 1,
        "operation": "select",
        "p": undefined,
      },
    },
    "state": "abcd",
    "status": "Active",
    "weight": 4,
  },
}
`;

exports[`the SelectorInstance class different arg-accepting selectors with the exact same name and args create different instances 1`] = `
{
  "@@selector-commonName-0-["a"]": {
    "className": "SelectorInstance",
    "observers": {},
    "sources": {},
    "state": "ab",
    "status": "Active",
    "weight": 1,
  },
  "@@selector-commonName-1-["a"]": {
    "className": "SelectorInstance",
    "observers": {},
    "sources": {},
    "state": "ac",
    "status": "Active",
    "weight": 1,
  },
}
`;
